flycheck.el

```diff
@@ -3035,19 +3122,26 @@ ERR is a Flycheck error.  MODE may be one of the following symbols:
 `lines'
      Return the line region.
 
-Otherwise signal an error."
+Otherwise signal an error.
+
+If ERR contains precise information about the region such as start
+and end columns, MODE is ignored."
   ;; Ignoring fields speeds up calls to `line-end-position' in
   ;; `flycheck-error-column-region' and `flycheck-error-line-region'.
-  (let ((inhibit-field-text-motion t))
-    (pcase mode
-      (`columns (or (flycheck-error-column-region err)
-                    (flycheck-error-line-region err)))
-      (`symbols (or (flycheck-error-thing-region 'symbol err)
-                    (flycheck-error-region-for-mode err 'columns)))
-      (`sexps (or (flycheck-error-thing-region 'sexp err)
-                  (flycheck-error-region-for-mode err 'columns)))
-      (`lines (flycheck-error-line-region err))
-      (_ (error "Invalid mode %S" mode)))))
+  (or (flycheck-error-region err)
+      (let ((inhibit-field-text-motion t)
+            (region
+             (or (flycheck-error-exact-region err)
+                 (pcase mode
+                   (`columns (or (flycheck-error-column-region err)
+                                 (flycheck-error-line-region err)))
+                   (`symbols (or (flycheck-error-thing-region 'symbol err)
+                                 (flycheck-error-region-for-mode err 'columns)))
+                   (`sexps (or (flycheck-error-thing-region 'sexp err)
+                               (flycheck-error-region-for-mode err 'columns)))
+                   (`lines (flycheck-error-line-region err))
+                   (_ (error "Invalid mode %S" mode))))))
+        (setf (flycheck-error-region err) region))))
```

Comment:
This is a bit of a change in behavior — it's legit to cache the region when we have a all 4 fields (beginning and end of line and column), but it cases like symbols or sexp, we don't currently cache the result.

So maybe we could make flycheck-error-exact-region set the `region` field (and return it), and there would be no more need for the final `setf`.